### ISSUE TITLE
Broadcast in Julia 1.11 requires `getindex`/`setindex!` to work on multiple integer indexing

### ISSUE THREAD

**Issue by charleskawczynski** [JuliaLang/julia#56295]
This didn't used to be the case in Julia 1.10, and this new requirement means that we cannot reserve this indexing for special behavior.

For example, the fix needed: https://github.com/CliMA/ClimaCore.jl/pull/2018

**Comment by charleskawczynski** [comment]
I'll try to make a reproducer at some point

**Comment by mbauman** [comment]
Interesting — these look like they're all non-Arrays who have axes and support get/setindex with `CartesianIndex` alone?

**Comment by charleskawczynski** [comment]
> they're all non-AbstractArrays who have axes and previously supported get/setindex with CartesianIndex alone?

Yeah, that's accurate. I put a fair bit of work into removing multiple integer indexing. We'd like to reserve linear indexing (i.e., multiple integer indexing) so that we can special-case fixes for https://github.com/JuliaLang/julia/issues/28126 (including the GPU-variant related issue resolution efforts: https://github.com/JuliaGPU/GPUArrays.jl/pull/454, https://github.com/JuliaGPU/GPUArrays.jl/pull/464).

**Comment by charleskawczynski** [comment]
Is there anything more that I can do to help this get fixed? (or is there any way we can work around this?)

**Comment by mbauman** [comment]
I mean, it’s not abundantly obvious to me that broadcast _should_ exclusively use one or the other — or _which one_ in the first place. I certainly couldn’t have told you that broadcast always generated `CartesianIndex`-based getindexes on 1.10. 

At a caller level, I have always implicitly assumed that the two should be perfectly equivalent. At the implementation level, `AbstractArray` considers n-ary `Int` to be the “canonical“ one. Having different behaviors between the two seems fraught for many algorithms, not just broadcast.

**Comment by charleskawczynski** [comment]
Ok, yes, I agree with you regarding design choices. We’re pretty heavily using broadcast in the CliMA org, and the 2x slowdown from Cartesian indexing is important enough that we need an alternative—whether that is some way to specialize on linear indexing, or roll our own custom version of broadcast.

I’ll look into the former, first, since that may be significantly less work.

**Comment by charleskawczynski** [comment]
It turns out that I was actually able to fix this with dispatch at a higher level.
